[regression/8.0.0] [Maui][Android] Signature/MAC verification failed

### Description

Since I updated to .NET8 RC2 SecureStorage.SetAsync does not work. It throws Javax.Crypto.AEADBadTagException with the following message: "Signature/MAC verification failed". 
Issue reproduced using Samsung A40 with Android 11. Case happens when backup rules exclude mauiessentials.xml file.

Below you can find stacktrace: 

```
  --- End of managed Android.Security.KeyStoreException stack trace ---
android.security.KeyStoreException: Signature/MAC verification failed
	at android.security.KeyStore.getKeyStoreException(KeyStore.java:1461)
	at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:186)
	at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
	at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:506)
	at javax.crypto.Cipher.doFinal(Cipher.java:2113)
	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decryptInternal(AndroidKeystoreAesGcm.java:118)
	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decrypt(AndroidKeystoreAesGcm.java:101)
	at com.google.crypto.tink.KeysetHandle.decrypt(KeysetHandle.java:993)
	at com.google.crypto.tink.KeysetHandle.readWithAssociatedData(KeysetHandle.java:878)
	at com.google.crypto.tink.KeysetHandle.read(KeysetHandle.java:859)
	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readMasterkeyDecryptAndParseKeyset(AndroidKeysetManager.java:378)
	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:298)
	at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:169)
	at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:130)
```

### Steps to Reproduce

Invoke Microsoft.Maui.Storage.SecureStorage.SetAsync

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.0-rc.2.9373

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.0-rc.1.9171

### Affected platforms

Android

### Affected platform versions

_No response_

### Did you find any workaround?

Yes

Issue doesn't happen when, in manifest, android:allowBackup is set to false as following:
<application android:allowBackup="false" ... >


### Relevant log output

_No response_

@figuerres I have the error on both debug and release

Also Allow Backup none doesn´t seem to be a workaround since all saved data get´s deleted after a restart of the app.

I´m having the same error when using any SecureStorage command:
 `[ViewRootImpl@5b71c8b[MainActivity]] ViewPostIme pointer 0
[ViewRootImpl@5b71c8b[MainActivity]] ViewPostIme pointer 1
[ViewRootImpl@5b71c8b[MainActivity]] onDisplayChanged oldDisplayState=2 newDisplayState=2
[DOTNET]    at Java.Interop.JniEnvironment.StaticMethods.CallStaticObjectMethod(JniObjectReference type, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/net7.0/JniEnvironment.g.cs:line 21452
[DOTNET]    at Java.Interop.JniPeerMembers.JniStaticMethods.InvokeObjectMethod(String encodedMember, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniStaticMethods.cs:line 165
[DOTNET]    at AndroidX.Security.Crypto.EncryptedSharedPreferences.Create(Context context, String fileName, MasterKey masterKey, PrefKeyEncryptionScheme prefKeyEncryptionScheme, PrefValueEncryptionScheme prefValueEncryptionScheme) in C:\a\_work\1\s\generated\androidx.security.security-crypto\obj\Release
et6.0-android\generated\src\AndroidX.Security.Crypto.EncryptedSharedPreferences.cs:line 227
[DOTNET]    at Microsoft.Maui.Storage.SecureStorageImplementation.GetEncryptedSharedPreferences() in D:\a\_work\1\s\src\Essentials\src\SecureStorage\SecureStorage.android.cs:line 93
[DOTNET]    at Microsoft.Maui.Storage.SecureStorageImplementation.<>c__DisplayClass6_0.<PlatformSetAsync>b__0() in D:\a\_work\1\s\src\Essentials\src\SecureStorage\SecureStorage.android.cs:line 53
[DOTNET]    at System.Threading.Tasks.Task.InnerInvoke()
[DOTNET]    at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj)
[DOTNET]    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
[DOTNET] --- End of stack trace from previous location ---
[DOTNET]    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
[DOTNET]    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
[DOTNET] --- End of stack trace from previous location ---
[DOTNET]    at Schaden_Digital.Views.PageLogin.Button_Clicked_1(Object sender, EventArgs e) in F:\Sourcecode\Online Gutachten\Online Gutachten\Views\Login\PageLogin.xaml.cs:line 77
[DOTNET]   --- End of managed Javax.Crypto.AEADBadTagException stack trace ---
[DOTNET] javax.crypto.AEADBadTagException
[DOTNET] 	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:632)
[DOTNET] 	at javax.crypto.Cipher.doFinal(Cipher.java:2114)
[DOTNET] 	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decryptInternal(AndroidKeystoreAesGcm.java:118)
[DOTNET] 	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decrypt(AndroidKeystoreAesGcm.java:101)
[DOTNET] 	at com.google.crypto.tink.KeysetHandle.decrypt(KeysetHandle.java:993)
[DOTNET] 	at com.google.crypto.tink.KeysetHandle.readWithAssociatedData(KeysetHandle.java:878)
[DOTNET] 	at com.google.crypto.tink.KeysetHandle.read(KeysetHandle.java:859)
[DOTNET] 	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readMasterkeyDecryptAndParseKeyset(AndroidKeysetManager.java:378)
[DOTNET] 	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:298)
[DOTNET] 	at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:169)
[DOTNET] 	at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:130)
[DOTNET] Caused by: android.security.KeyStoreException: Signature/MAC verification failed (internal Keystore code: -30 message: system/security/keystore2/src/operation.rs:850: KeystoreOperation::finish
[DOTNET] 
[DOTNET] Caused by:
[DOTNET]     0: system/security/keystore2/src/operation.rs:426: Finish failed.
[DOTNET]     1: Error::Km(r#VERIFICATION_FAILED)) (public error code: 10 internal Keystore code: -30)
[DOTNET] 	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:435)
[DOTNET] 	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
[DOTNET] 	at android.security.KeyStoreOperation.finish(KeyStoreOperation.java:128)
[DOTNET] 	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
[DOTNET] 	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
[DOTNET] 	at android.security.keystore2.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:396)
[DOTNET] 	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
[DOTNET] 	... 10 more
[DOTNET] Exception of type 'Javax.Crypto.AEADBadTagException' was thrown.
[ViewRootImpl@5b71c8b[MainActivity]] onDisplayChanged oldDisplayState=2 newDisplayState=2`

--- a/src/Essentials/src/SecureStorage/SecureStorage.android.cs
+++ b/src/Essentials/src/SecureStorage/SecureStorage.android.cs
@@ -3,6 +3,7 @@ using Android.App;
 using Android.Content;
 using AndroidX.Security.Crypto;
 using Java.Security;
+using Javax.Crypto;
 using Xamarin.Google.Crypto.Tink.Shaded.Protobuf;
 
 namespace Microsoft.Maui.Storage
@@ -80,31 +81,64 @@ namespace Microsoft.Maui.Storage
 			editor?.Clear()?.Apply();
 		}
 
+		static void DeleteSharedPreferences()
+		{
+			// Get the alias used to store the encrypted values
+			var alias = $"{ApplicationModel.AppInfo.Current.PackageName}.microsoft.maui.essentials.preferences";
+			// Open an editor to the preferences we can clear
+			var editPreferences = Application.Context.GetSharedPreferences(alias, FileCreationMode.Private).Edit();
+			// Commit is synchronous here so we can be sure it's done before trying to create the encrypted preferences again
+			editPreferences?.Clear()?.Commit();
+		}
+
 		ISharedPreferences GetEncryptedSharedPreferences()
 		{
 			try
 			{
-				var context = Application.Context;
-
-				var prefsMainKey = new MasterKey.Builder(context, Alias)
-					.SetKeyScheme(MasterKey.KeyScheme.Aes256Gcm)
-					.Build();
-
-				return EncryptedSharedPreferences.Create(
-					context,
-					Alias,
-					prefsMainKey,
-					EncryptedSharedPreferences.PrefKeyEncryptionScheme.Aes256Siv,
-					EncryptedSharedPreferences.PrefValueEncryptionScheme.Aes256Gcm);
+				return CreateEncryptedSharedPreferences();
 			}
-			catch (InvalidProtocolBufferException)
+			catch (System.Exception ex) 
+			when (ex is InvalidProtocolBufferException or Android.Security.KeyStoreException or KeyStoreException or BadPaddingException)
 			{
-				// TODO: Use Logger here?
-				System.Diagnostics.Debug.WriteLine(
-					"Unable get encrypted shared preferences, which is likely due to an app uninstall. Removing all keys and returning null.");
-				PlatformRemoveAll();
+				// If we encounter any of these exceptions, it's likely due to a corrupt key or bad migration between devices
+				// There isn't much to do at this point except try to delete the shared preferences so we can recreate them
+				try
+				{
+					System.Diagnostics.Debug.WriteLine(
+						"Unable get encrypted shared preferences, which is likely due to corrupt encryption key or bad app cache backup/restore. Removing all keys and returning null.");
+					System.Diagnostics.Debug.WriteLine(ex);
+
+					// Delete the shared preferences
+					DeleteSharedPreferences();
+
+					// Try to return a new instance now that we've deleted the old
+					return CreateEncryptedSharedPreferences();
+				}
+				catch (System.Exception ex2)
+				{
+					// If we still can't create things, we'll have to give up and return null
+					// TODO: Use Logger here?
+					System.Diagnostics.Debug.WriteLine("Still unable to create encrypted shared preferences after attempting to deleting them. Returning null.");
+					System.Diagnostics.Debug.WriteLine(ex2);
+				}
 				return null;
 			}
+		}
+
+		ISharedPreferences CreateEncryptedSharedPreferences()
+		{
+			var context = Application.Context;
+
+			var prefsMainKey = new MasterKey.Builder(context, Alias)
+				.SetKeyScheme(MasterKey.KeyScheme.Aes256Gcm)
+				.Build();
+
+			return EncryptedSharedPreferences.Create(
+				context,
+				Alias,
+				prefsMainKey,
+				EncryptedSharedPreferences.PrefKeyEncryptionScheme.Aes256Siv,
+				EncryptedSharedPreferences.PrefValueEncryptionScheme.Aes256Gcm);
 		}
 	}
 }

--- a/src/Essentials/src/SecureStorage/SecureStorage.android.cs
+++ b/src/Essentials/src/SecureStorage/SecureStorage.android.cs
@@ -83,10 +83,8 @@ namespace Microsoft.Maui.Storage
 
 		static void DeleteSharedPreferences()
 		{
-			// Get the alias used to store the encrypted values
-			var alias = $"{ApplicationModel.AppInfo.Current.PackageName}.microsoft.maui.essentials.preferences";
-			// Open an editor to the preferences we can clear
-			var editPreferences = Application.Context.GetSharedPreferences(alias, FileCreationMode.Private).Edit();
+			// Open an editor to the preferences we can clear, using the alias for storing encrypted values
+			var editPreferences = Application.Context.GetSharedPreferences(Alias, FileCreationMode.Private).Edit();
 			// Commit is synchronous here so we can be sure it's done before trying to create the encrypted preferences again
 			editPreferences?.Clear()?.Commit();
 		}

